### ISSUE TITLE
Unreliable internet connection from pods

### ISSUE THREAD

**Issue by fungiboletus** [k3s-io/k3s#5220]
**Environmental Info:**
K3s Version: `v1.22.7+k3s1`

Node(s) CPU architecture, OS, and Version:
`Linux k3s-host-0 5.13.0-30-generic #33~20.04.1-Ubuntu SMP Mon Feb 7 14:25:10 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

Cluster Configuration:
1 single host.

**Describe the bug:**
Internet connection is not reliable inside the pods. It works randomly.

**Steps To Reproduce:**

Install K3S on Ubuntu 20.04 LTS with the following command: `curl -sfL https://get.k3s.io | sh -`

In my case, the server is a random bare metal server from hetzner.

Install tcptraceroute on the host, do a traceroute to a random reliable service, and notice it works.
`tcptraceroute google.com 80` for example.

Do the same inside a pod:

```
kubectl run test-pod --rm -i --tty --image alpine -- /bin/sh`
# wait
# Install tcptraceroute, may require multiple tentatives
apk add tcptraceroute
tcptraceroute google.com 80
```

Notice that it works randomly, but not always.  It's always the last server that does not respond. All intermediary routers do respond.

**Expected behaviour:**

Internet connection should be reliable.

**Actual behaviour:**

Internet connection is not reliable and the tcp packets do not seem to be routed correctly.

**Additional context / logs:**

I installed Docker temporarily and the internet connection works correctly inside the containers. I removed it after testing, don't blame docker.

I also tried the previous version (1.22.6) and the latest version without success.

My installation is very vanilla and I have a similar setup in another VM from another cloud provider that has no issues. I know the problem doesn't appear everywhere, but I have no idea why it appears on my setup and I think it may be present in other similar installations.

**Comment by clementnuss** [comment]
😮 I just lost 2 hours trying to figure out why my pods couldn't reach internet anymore.

I reverted back to `v1.22.6+k3s1` which instantly reverted the issue.

for what it's worth, I've been experimenting a bit during the issue, and noticed that `udp` traffic was working correctly.
moreover, with `iptables` I noticed that TCP connections were established with the internet peers, and the answer came back, but it's as if the container never processed the internet peer answer. masquerading looked good though, and the fact that `udp` is working would also indicate another issue.

this definitely is a major bug in 1.22.7 😬

**Comment by fungiboletus** [comment]
The revers didn't work for me. What is your kernel version by any chance ?

**Comment by brandond** [comment]
cc @manuelbuil perhaps something on the flannel side?

**Comment by clementnuss** [comment]
kernel version: `5.11.0-49-generic` on an Ubuntu 21.04 node and `5.10.0-9-cloud-amd64` on a Debian 11 node

I'm using `Cilium` and therefore running my k3s with the following options:

```
ExecStart=/usr/local/bin/k3s \
    server \
    --flannel-backend=none \
    --disable servicelb \
    --disable traefik \
    --disable local-storage \
    --disable-kube-proxy \
```

Not sure how this could have an impact. would be interesting to know @fungiboletus what settings you use

**Comment by brandond** [comment]
Hmm, if you're using your own CNI, perhaps the same issue as this then:
* https://github.com/k3s-io/k3s/issues/5188#issuecomment-1059072020

**Comment by fungiboletus** [comment]
On my side I only use the defaults settings.

**Comment by brandond** [comment]
@fungiboletus Does disabling the network policy controller have any effect on your issue?

**Comment by manuelbuil** [comment]
> kernel version: `5.11.0-49-generic` on an Ubuntu 21.04 node and `5.10.0-9-cloud-amd64` on a Debian 11 node
> 
> I'm using `Cilium` and therefore running my k3s with the following options:
> 
> ```
> ExecStart=/usr/local/bin/k3s \
>     server \
>     --flannel-backend=none \
>     --disable servicelb \
>     --disable traefik \
>     --disable local-storage \
>     --disable-kube-proxy \
> ```
> 
> Not sure how this could have an impact. would be interesting to know @fungiboletus what settings you use

As @brandond was saying, please disable the network policy controller. As stated in the docs: "Run K3s with --flannel-backend=none and install your CNI of choice. Most CNI plugins come with their own network policy engine, so it is recommended to set --disable-network-policy as well to avoid conflicts." https://rancher.com/docs/k3s/latest/en/installation/network-options/#custom-cni

**Comment by manuelbuil** [comment]
@fungiboletus Could you please share the output of the following commands:
```
ip a
ip r
ip rule
cat /proc/sys/net/ipv4/conf/all/forwarding
cat /proc/sys/net/ipv4/conf/eth0/forwarding (or replace eth0 with whatever is your main interface)
```

**Comment by clementnuss** [comment]
> > kernel version: `5.11.0-49-generic` on an Ubuntu 21.04 node and `5.10.0-9-cloud-amd64` on a Debian 11 node
> > I'm using `Cilium` and therefore running my k3s with the following options:
> > ```
> > ExecStart=/usr/local/bin/k3s \
> >     server \
> >     --flannel-backend=none \
> >     --disable servicelb \
> >     --disable traefik \
> >     --disable local-storage \
> >     --disable-kube-proxy \
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Not sure how this could have an impact. would be interesting to know @fungiboletus what settings you use
> 
> @[clementnuss](https://github.com/clementnuss) As @brandond was saying, please disable the network policy controller. As stated in the docs: "Run K3s with --flannel-backend=none and install your CNI of choice. Most CNI plugins come with their own network policy engine, so it is recommended to set --disable-network-policy as well to avoid conflicts." https://rancher.com/docs/k3s/latest/en/installation/network-options/#custom-cni

~~This probably isn't an issue with NetPol: I've done the upgrade again to `1.22.7` and I'm again in the failure mode.~~

you were right, adding `--disable-network-policy` solved my problem! there is one major catch, though:
all pods have to be restarted, for the solution to be effective (and obviously the `k3s` service before that)

thanks a lot 👏🏻 🎉

**Comment by manuelbuil** [comment]
> > > kernel version: `5.11.0-49-generic` on an Ubuntu 21.04 node and `5.10.0-9-cloud-amd64` on a Debian 11 node
> > > I'm using `Cilium` and therefore running my k3s with the following options:
> > > ```
> > > ExecStart=/usr/local/bin/k3s \
> > >     server \
> > >     --flannel-backend=none \
> > >     --disable servicelb \
> > >     --disable traefik \
> > >     --disable local-storage \
> > >     --disable-kube-proxy \
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Not sure how this could have an impact. would be interesting to know @fungiboletus what settings you use
> > 
> > 
> > @[clementnuss](https://github.com/clementnuss) As @brandond was saying, please disable the network policy controller. As stated in the docs: "Run K3s with --flannel-backend=none and install your CNI of choice. Most CNI plugins come with their own network policy engine, so it is recommended to set --disable-network-policy as well to avoid conflicts." https://rancher.com/docs/k3s/latest/en/installation/network-options/#custom-cni
> 
> ~This probably isn't an issue with NetPol: I've done the upgrade again to `1.22.7` and I'm again in the failure mode.~
> 
> you were right, adding `--disable-network-policy` solved my problem! there is one major catch, though: all pods have to be restarted, for the solution to be effective (and obviously the `k3s` service before that)
> 
> thanks a lot 👏🏻 tada

Yay!!! :partying_face:

**Comment by fungiboletus** [comment]
@manuelbuil  Thanks for following up. I uploaded the outputs [on a gist](https://gist.github.com/fungiboletus/d1335872d3f8479ba51461c2445ed482).

It's while running k3s with the `--disable-network-policy` option, which doesn't fix the issue for me.

I also uploaded another log file where I start an alpine pod and run tcptraceroute. Installing tcptraceroute took 2 tentatives, and the tcptraceroute fails randomly to reach destination as you may see.

**Comment by manuelbuil** [comment]
> @manuelbuil Thanks for following up. I uploaded the outputs [on a gist](https://gist.github.com/fungiboletus/d1335872d3f8479ba51461c2445ed482).
> 
> It's while running k3s with the `--disable-network-policy` option, which doesn't fix the issue for me.
> 
> I also uploaded another log file where I start an alpine pod and run tcptraceroute. Installing tcptraceroute took 2 tentatives, and the tcptraceroute fails randomly to reach destination as you may see.

Thanks, have you noticed that you have the same ip twice in your main interface but with different mask?

```
2: enp41s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether a8:a1:59:8b:33:b3 brd ff:ff:ff:ff:ff:ff
    inet 135.181.20.157/32 scope global enp41s0
       valid_lft forever preferred_lft forever
    inet 135.181.20.157/26 brd 135.181.20.191 scope global enp41s0
       valid_lft forever preferred_lft forever
    inet6 2a01:4f9:4b:44c9::2/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::aaa1:59ff:fe8b:33b3/64 scope link 
       valid_lft forever preferred_lft forever
```
As it is the same ip, it should not matter but still.

Could you try to `ping 8.8.8.8` from the alpine pod? I'd like to see if it works. traceroute is not the most reliable tool ;)

I would also like to see the output of:

```
kubectl get pods -A -o wide
```

**Comment by fungiboletus** [comment]
Thanks. Indeed, I had both a static IP configuration and DHCP enabled. I fixed that and also disabled IPv6 (doesn't need it to debug).

A created [a new gist with the updated outputs and the list of pods](https://gist.github.com/fungiboletus/dd2ce1d5fd9a9c75a2d204f254466784).

Ping is working.

**Comment by manuelbuil** [comment]
Thanks. When you say: `Internet connection is not reliable and the tcp packets do not seem to be routed correctly.`, it means tcptraceroute fails, right? Or are you seeing any other problem?

**Comment by fungiboletus** [comment]
The main problem is that HTTP(s) requests fail or are very slow. tcptraceroute is nice to illustrate the problem, but curl also fail randomly for example.

**Comment by manuelbuil** [comment]
Ok, so for example `curl -I http://www.google.com` fails, right?

Can you check if you get any output on:
```
kubectl get networkpolicy -A
```

**Comment by fungiboletus** [comment]
Yes this command fail (I just configured the timeout to not wait 5 minutes like by default):

```
$ curl --connect-timeout 30 -I http://www.google.com
curl: (28) Failed to connect to www.google.com port 80 after 15202 ms: Operation timed out
```

I don't have network policy, the command returns `No resources found`.

**Comment by fungiboletus** [comment]
I found the issue ! Thanks a lot for the help @manuelbuil The problem was not from k3s but my installation, which was likely the case but I really spend time trying to remove everything else in my system.

My Hetzner server used a default firewall configuration which is simply broken for k3s and probably many other things. It blocks incoming ack tcp packets on ports below 32768. I edited the firewal rule to allow all ports and it's working again.

Broken firewall:
![image](https://user-images.githubusercontent.com/45740/157444128-7efa3921-59a1-4d9a-8a5f-4a07eca521de.png)
